cleangoldenami/main.go

```diff
@@ -58,23 +59,70 @@ func main() {
+				numOfImagesToDelete := desiredImageCapacity - (quota - imageCount)
+
+				arm64Images := filterImages(images, "arm64", "rhel-arm64")
+				if err != nil {
+					log.Fatalf("error fetching rhel arm64 images for region %v: %v", regionName, err)
+				}
+
+				legacyx86Images := filterImages(images, "x86_64", "legacy-x86_64")
+				if err != nil {
+					log.Fatalf("error fetching legacy x86_64 images for region %v: %v", regionName, err)
+				}
+
+				x86Images := filterImages(images, "x86_64", "rhel-x86_64")
 				if err != nil {
-					log.Fatalf("error determining which image to delete in region %v: %v", regionName, err)
+					log.Fatalf("error fetching rhel x86_64 images for region %v: %v", regionName, err)
+				}
+				var imagesToDelete []ec2Types.Image
+
+				for !imageDeletionCheck(imagesToDelete, numOfImagesToDelete, arm64Images, legacyx86Images, x86Images) {
+					mostPopulatedImagesByType, mostPopulatedImageType := getMostPopulatedImageType(arm64Images, legacyx86Images, x86Images)
+
+					ImageToDeleteIndex, ImageToDelete, err := getOldestImage(mostPopulatedImagesByType)
+					if err != nil {
+						log.Fatalf("error determining which image to delete in region %v: %v", regionName, err)
+					}
+
+					if ImageToDeleteIndex >= 0 {
+						imagesToDelete = append(imagesToDelete, ImageToDelete)
+
+						switch mostPopulatedImageType {
+						case "arm64Images":
+							arm64Images = append(arm64Images[:ImageToDeleteIndex], arm64Images[ImageToDeleteIndex+1:]...)
+						case "legacyx86Images":
+							legacyx86Images = append(legacyx86Images[:ImageToDeleteIndex], legacyx86Images[ImageToDeleteIndex+1:]...)
+						case "x86Images":
+							x86Images = append(x86Images[:ImageToDeleteIndex], x86Images[ImageToDeleteIndex+1:]...)
+						}
+					}
 				}
 				if *dryRun {
-					fmt.Printf("Region %v is at quota (%v) - would delete %v\n", regionName, quota, *imageToDelete.ImageId)
+					for _, image := range imagesToDelete {
+						for i, t := range image.Tags {
+							if *t.Key == "version" {
+								fmt.Printf("Region %v is at quota (%v) - would delete %v (%v)\n", regionName, quota, *image.ImageId, *image.Tags[i].Value)
+							}
+						}
+					}
 				} else {
-					err := deregisterImage(ec2Client, imageToDelete)
-					if err != nil {
-						log.Fatalf("error deregistering image %v in region %v: %v", *imageToDelete.ImageId, regionName, err)
+					for _, image := range imagesToDelete {
+						for i, t := range image.Tags {
+							if *t.Key == "version" {
+								err = deregisterImage(ec2Client, image)
+								if err != nil {
+									fmt.Printf("error deregistering image %v (%v) in region %v: %v\n", *image.ImageId, *image.Tags[i].Value, regionName, err)
+								}
+								fmt.Printf("successfully deregistered image %v (%v) in region %v\n", *image.ImageId, *image.Tags[i].Value, regionName)
+							}
+						}
 					}
-					fmt.Printf("successfully deregistered image %v in region %v", *imageToDelete.ImageId, regionName)
 				}
```

Comment:
Nit: you might consider moving the outer if-statement into the innermost loop in order to reduce repetition, e.g.,
```golang
					for _, image := range imagesToDelete {
						for i, t := range image.Tags {
							if *t.Key == "version" {
                                if !*dryRun {
								  err = deregisterImage(ec2Client, image)
								  if err != nil {
									  fmt.Printf("error deregistering image %v (%v) in region %v: %v\n", *image.ImageId, *image.Tags[i].Value, regionName, err)
								  }
								  fmt.Printf("successfully deregistered image %v (%v) in region %v\n", *image.ImageId, *image.Tags[i].Value, regionName)
                              } else { 
                                  fmt.Printf("Region %v is at quota (%v) - would delete %v (%v)\n", regionName, quota, *image.ImageId, *image.Tags[i].Value) 
                              }
							}
						}
					}
```